### ISSUE TITLE
Global steps vs epochs

### ISSUE THREAD

**Issue by tomsmith82** [Kyubyong/tacotron#49]
Can someone explain to me the difference between epochs and global steps? I'm training a few thousand epochs, but the global steps stay low.

**Comment by jaron** [comment]
An epoch is one full pass through the training set, so that every sample gets the chance to be seen. But because it's impractical to hold the entire data set in memory, the convention is to use batching, an iterative process where each batch is a smaller sample of training data used to improve of your model.

Global steps refer to the number of batches seen by the graph.

Say you set the num_epochs parameter to 100, start training and checkpoint your model. If you have one batch per epoch, your global step will be 100. (In practice it'll be more, because you're training on batches).

Then you decide to do some more processing, so you restore the model and run another 100 epochs. Now your global step is 200, but if you look in logdir the epoch number will be a maximum of 100. Epochs are local to a training run, whereas the global step records the total number of iterations. This is useful because the knowing the amount of training a model has undergone can be used to tweak hyper-parameters, like the decaying the learning rate over time to encourage convergence, as mentioned on page 5 of the [Tacotron paper](https://arxiv.org/pdf/1703.10135.pdf).

The global step should always be the same or higher than your epoch count. If you're seeing a low number of global steps, there might be a training error. I'd suspect you probably haven't completed many epochs.

**Comment by reiinakano** [comment]
@jaron Does it make sense to keep the number of epochs local to a training run? I actually modified the code a bit so it records the number of *global* epochs.

**Comment by jaron** [comment]
@reiinakano Personally, I'd tend to stick to the usual naming convention, so it's easier for others to understand what you're doing. For instance, Tensorboard uses global steps as its measure for  training progress.
